scene/2d/audio_stream_player_2d.cpp

```diff
@@ -164,28 +175,15 @@ void AudioStreamPlayer2D::_update_panning() {
 		volume_vector.write[0] = AudioFrame(l, r) * multiplier;
 	}
 
-	AudioServer::get_singleton()->set_playback_bus_exclusive(stream_playback, _get_actual_bus(), volume_vector);
+	MutexLock lock(playback_list_mutex);
+	for (const Ref<AudioStreamPlayback> &playback : stream_playbacks) {
+		AudioServer::get_singleton()->set_playback_bus_exclusive(playback, _get_actual_bus(), volume_vector);
+	}
 }
 
 void AudioStreamPlayer2D::set_stream(Ref<AudioStream> p_stream) {
-	if (stream_playback.is_valid()) {
-		stop();
-	}
-
-	stream_playback.unref();
-	stream.unref();
-	if (p_stream.is_valid()) {
-		stream_playback = p_stream->instance_playback();
```

Comment:
I still think there is a big problem with this approach, and is the fact that for generating sound from the main thread via AudioStreamGenerator, you need to access and keep the playback object. This will be also likely needed if something like VoIP is implemented, as you need to push the sound that comes from the network via a ringbuffer.

I think we could work around this by using the following logic:

* If the stream is polyphonic, we do things as-is, current code-path is used.
* If the stream is monophonic, we request and cache the AudioStreamPlayback and use a single sized stream_playback, and never free it until the stream is changed or removed.

Reply:
Implemented this behavior. I also went a step farther and added a warning. Let me know if you'd like to remove this but I don't think anyone should ever be calling play() if the player is already playing a non-polyphonic stream. This would be a source of bugs for AudioStreamGenerators and would at best cause disruptions in audio playback during fade-out/fade-in for AudioStreamPlaybackMicrophone. Those are the two non-polyphonic streams in core I think, so I thought I'd add this warning. It won't spam in exported builds though because I put it inside an `#ifdef TOOLS_ENABLED`. Let me know if you want me to remove this when I squash the commits though. It would be simple enough to do so but unless there's a use-case for attempting polyphonic playback on non-polyphonic streams I'd prefer to keep it.